jax/interpreters/ad.py

```diff
@@ -506,34 +509,37 @@ def defbilinear_broadcasting(bcast, prim, lhs_rule, rhs_rule):
 
 def bilinear_transpose(lhs_rule, rhs_rule, cotangent, x, y, **kwargs):
   assert is_undefined_primal(x) ^ is_undefined_primal(y)
+  if type(cotangent) is Zero:
+    return Zero
   if is_undefined_primal(x):
-    out = zero if cotangent is zero else lhs_rule(cotangent, y, **kwargs)
-    return out, None
+    out = lhs_rule(cotangent, y, **kwargs)
+    return Zero if out is Zero else (out, None)
   else:
-    out = zero if cotangent is zero else rhs_rule(cotangent, x, **kwargs)
-    return None, out
+    out = rhs_rule(cotangent, x, **kwargs)
+    return Zero if out is Zero else (None, out)
 
 
 def defjvp_zero(primitive):
   assert isinstance(primitive, Primitive)
   primitive_jvps[primitive] = partial(zero_jvp, primitive)
 
 def zero_jvp(primitive, primals, tangents, **params):
-  return primitive.bind(*primals, **params), zero
+  r = primitive.bind(*primals, **params)
+  return r, Zero.from_value(r)
 
 
-deflinear(zeros_like_p, lambda t: [zero])
+deflinear(zeros_like_p, lambda t: [Zero.from_value(t)])
 deflinear(core.identity_p, lambda t: (t,))
 deflinear(add_jaxvals_p, lambda t: (t, t))
 
 def instantiate_zeros(example, tangent):
-  if tangent is zero:
+  if type(tangent) is Zero:
     return zeros_like_jaxval(example)
```

Comment:
We don't need to pass `example` into this function anymore AIUI, though it might be annoying to update all the call sites.

Instead, perhaps we should add a check that `example` is consistent with `tangent.aval` when `type(tangent) is Zero`? WDYT?

Reply:
Oh that's really nice, I didn't realize that. Let me update the call sites!